### PR TITLE
Add rating widget

### LINKED ISSUE #2359: Add 'rating' appearance for range type
As described in https://github.com/XLSForm/pyxform/issues/145.

The widget should validate that the `step` is 1 and that `min` is 1. There should be `max` stars displayed with equal spacing. If the stars don't fit in the device width, they should wrap onto additional lines.

When the user taps on an empty (white) star, the stars up to and including that star should be filled in yellow and the stored value should be updated accordingly. When the user taps on a yellow star, if there are stars to the right of it that are also yellow, those stars should become empty/white and the stored value should be updated accordingly. If there are no yellow stars to the right of the yellow star tapped, the state of the widget should remain unchanged.

If no default is specified, no stars should be selected initially and the question value should be blank. If a default is specified, that number of stars should be selected initially. Clearing the widget should clear all the stars.

Requirement and constraints should work as they do with other question types.

### PR DESCRIPTION
**Pull request by akshay-ap**
Closes #2359 

#### What has been done to verify that this works as intended?
I have tested the behaviour on Moto M. With the form containing `range` widget of apperance `rating`.

#### Why is this the best possible solution? Were any other approaches considered?
I have used `RatingBar` provided by android.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Users can add appreance of type `rating`  in `range` widget.

#### Do we need any specific form for testing your changes? If so, please attach one.
Yes. I will share the form in the thread.
#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by akshay-ap** [comment]
I have used the following form:
[Form.zip](https://github.com/opendatakit/collect/files/2421160/Form.zip)

To use `max` in `parameters`, there seems to be a need to make changes to `javarosa` library.
Currently, the number of stars is defined by `end` in `parameters` column.
Known issue : Stars go out of screen due to limitation of screen size.

File: collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java [hunk 1]
```diff
@@ -277,21 +273,28 @@ public static QuestionWidget createWidgetFromPrompt(FormEntryPrompt fep, Context
...
+                    questionWidget = new RatingWidget(context, fep);
+                } else {
+                    switch (fep.getDataType()) {
+                        case Constants.DATATYPE_INTEGER:
+                            questionWidget = new RangeIntegerWidget(context, fep);
+                            break;
+                        case Constants.DATATYPE_DECIMAL:
+                            questionWidget = new RangeDecimalWidget(context, fep);
+                            break;
+                        default:
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java:291]: Why did you add this default case? It should not be necessary.

File: collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java [hunk 2]
```diff
@@ -277,21 +273,28 @@ public static QuestionWidget createWidgetFromPrompt(FormEntryPrompt fep, Context
...
+                            break;
+                    }
+                }
+                    break;
                     default:
                         questionWidget = new StringWidget(context, fep, readOnlyOverride);
                         break;
-                }
-                break;
-            default:
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java]: This default case is important, though, because it's a catch-all for unrecognized types. Why did you remove it?

**Comment by lognaturel** [comment]
Thanks for taking this on, @akshay-ap! It doesn't quite match the description because tapping on a specific star doesn't fill or empty that star. That's important to get right because it's how ratings in other apps work.

If I tap to set a particular rating, swipe forward and then swipe back, the value is not saved.

> To use max in parameters, there seems to be a need to make changes to javarosa library.
Currently, the number of stars is defined by end in parameters column.

`end` is the correct value to use, thanks.

> Known issue : Stars go out of screen due to limitation of screen size.

This is fine and expected.

File: collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java (hunk 2 above)
Comment (akshay-ap) [reply collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java]: Looks like I made some unintentional changes. Will correct it.

**Comment by akshay-ap** [comment]
@lognaturel How about using `GridLayout` (Each grid location containing ImageView) to make a rating appearance instead of using `RatingBar`? It will solve the problem of screen size limitation by placing the stars in rows and columns.

File: collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java (hunk 1 above)
Comment (akshay-ap) [reply collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java:291]: The default case is already present in code.

**Comment by lognaturel** [comment]
The values are now saved which is great! 

> It will solve the problem of screen size limitation by placing the stars in rows and columns.

This issue isn't a problem to me. We can document the limitation but I don't think most people would want more than a few stars anyway.

The problem I see is that tapping on a star has a result that is off by one from what I expect. The video at https://www.youtube.com/watch?v=UbsBH180H34 shows what I would expect. For example, tapping on the first star fills just that star. When I run this branch, tapping on the first star fills the two first stars on Android 5.1.1. Do you see that?

**Comment by akshay-ap** [comment]
> For example, tapping on the first star fills just that star. When I run this branch, tapping on the first star fills the two first stars on Android 5.1.1. Do you see that?

Yes. I am facing similar behavior on android 7.0.

**Comment by akshay-ap** [comment]
@lognaturel I think `RatingBar` contains some bug. What would you suggest?

**Comment by shobhitagarwal1612** [comment]
Hello @akshay-ap 
It seems like the problem is not with the widget but the way we are using it. Since the precision of the rating bar or the step size is `1.0`, so it will always return only integer values.
The reason why it was behaving erratically was that we were clicking on the right half of the star which gets rounded off to the next biggest integer. Just try clicking on the left half of the star (for fun!) and it would select only that part. 
So, to solve this problem, just set the `stepSize` to `0.5` as then it would not round off >0.5 values to the next integer

**Comment by shobhitagarwal1612** [comment]
Also, there is no way to clear the answer! (long press event)

**Comment by lognaturel** [comment]
@shobhitagarwal1612, using a `stepSize` of `0.5` would make it possible to save values that end in `.5`, right? That would not be desirable.

@akshay-ap, have you read through https://issuetracker.google.com/issues/37114040? Seems some API levels are affected by a bug. I've also found some other interesting resources by searching for "ratingbar android stepsize". Would be good to read through those and see if it gives any ideas.

**Comment by shobhitagarwal1612** [comment]
> @shobhitagarwal1612, using a stepSize of 0.5 would make it possible to save values that end in .5, right? That would not be desirable

How about if we set the step size as 0.5 but when the value gets changed, only use the floor value of the float as the answer and also setting the current value of rating bar equal to that?

**Comment by akshay-ap** [comment]
@lognaturel Please review.
</img>
(with 100 stars)

File: collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            imageButton.setImageResource(R.drawable.ic_star_border);
+            imageButton.setId(total);
+            imageButton.setPadding(0, 0, 0, 0);
+            imageButton.setBackground(null);
+            imageButton.setOnClickListener(view -> {
+                int position = view.getId();
+                for (int i = 0; i <= position; i++) {
+                    ImageButton button = (ImageButton) gridLayout.getChildAt(i);
+                    button.setImageResource(R.drawable.ic_star);
+                    answer = position + 1;
```
Comment (shobhitagarwal1612) [inline collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java]: This can be moved out of the `for` loop

File: collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                }
+
+                for (int i = position + 1; i < numberOfStarsToShow; i++) {
+                    ImageButton button = (ImageButton) gridLayout.getChildAt(i);
+                    button.setImageResource(R.drawable.ic_star_border);
```
Comment (shobhitagarwal1612) [inline collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java]: Both of the `for` loops can be combined together to simplify the logic. 
Something like:
```
button.setImageResource(i <= position ? R.drawable.ic_star : R.drawable.ic_star_border);
```

**Review by shobhitagarwal1612 — APPROVED** [review]

**Comment by kkrawczyk123** [comment]
I have noticed only one small issue on Android 8.1, Nexus 5X: when stars don’t fit to the one line (9 or 10 stars) I rotate a screen they are still wrapped, like on attached screenshot:
screenshot_20181031-091030
When I move backward or forward, stars fit to the screen like:
screenshot_20181031-091038
and when I rotate back they are not wrapped as they should vertically:
screenshot_20181031-091058
 I have to move backward or forward again so they will be wrapped again.

@opendatakit-bot unlabel "needs testing"

**Comment by lognaturel** [comment]
Thanks, all! I don't think the rotation issue needs to be addressed now since it is so specific.

At our last TSC meeting, @smoyth, @adamvert and @shobhitagarwal1612 expressed interest in being involved in UI/UX review, thank you! This seems like a nice, simple place to start. You can see in the previous conversation that we went back and forth on how important the wrapping was and ultimately decided it was important since we do support such a broad range of screen widths.

The questions I see that you may want to weigh on:
- is the scaling and spacing of the stars as you expect? It matches the default Android widget.
- the filled stars are currently black. Should they be yellow? The default highlight color (blue used in e.g. for settings icons)? Some other color?

Is there anything else you think should be considered from a UI/UX perspective?

**Comment by smoyte** [comment]
I agree needs to wrap. Spacing looks fine. Stars are almost always yellow or orange:

image
image
image

I would say stick with that to be safe. Black stars, blue stars, etc. may be special cultural symbols in some contexts. The same may be true for yellow but at least there we're relying on a widely used standard.

Orange or orange-y yellow may be better than yellow unless we want to use a darker border around the star (a la Amazon). Otherwise contrast with the white BG will be hard.

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
Thank you, @smoyth! I'm going to break with the usual process a little bit because this PR has been open for a long time and I remember that @akshay-ap was interested in [Hacktoberfest](https://hacktoberfest.digitalocean.com/) and the deadline is tonight. So I will merge this and open an issue to change the star color which is great for @akshay-ap or anyone else to take on. If @adamvert or @shobhitagarwal1612 have other feedback, please also file issues.

Again, typically we'd wait to address all feedback in the PR but given the circumstances I think this is an ok way to proceed.
